Framework/DataHandling/src/MaskDetectors.cpp

```diff
@@ -495,5 +500,65 @@ void MaskDetectors::appendToIndexListFromMaskWS(
   tmp_index.swap(indexList);
 } // appendToIndexListFromWS
 
+/**
+ * Append the detector IDs of detectors found recursively in the list of
+ * components.
+ *
+ * @param detectorList :: An existing list of detector IDs
+ * @param componentList :: List of component names
+ * @param WS :: Workspace instrument of which to use
+ */
+void MaskDetectors::appendToDetectorListFromComponentList(
+    std::vector<detid_t> &detectorList,
+    const std::vector<std::string> &componentList,
+    const API::MatrixWorkspace_const_sptr WS) {
+  const auto instrument = WS->getInstrument();
+  if (!instrument) {
+    g_log.error()
+        << "No instrument in input workspace. Ignoring ComponentList\n";
+    return;
+  }
+  std::set<detid_t> detectorIDs;
+  for (const auto &compName : componentList) {
+    std::vector<boost::shared_ptr<const IComponent>> components;
+    if (compName.find('/') != std::string::npos) {
+      auto component = instrument->getComponentByName(compName);
+      if (component) {
+        components.emplace_back(component);
+      }
+    } else {
+      components = instrument->getAllComponentsWithName(compName);
+    }
+    if (components.empty()) {
+      g_log.error() << "Component " << compName
+                    << " not found in input workspace.\n";
+      continue;
+    }
+    for (const auto &comp : components) {
+      const auto assembly =
+          boost::dynamic_pointer_cast<const ICompAssembly>(comp);
+      const auto detector = boost::dynamic_pointer_cast<const IDetector>(comp);
+      if (detector) {
+        detectorIDs.emplace(detector->getID());
+      } else if (assembly) {
+        std::vector<IComponent_const_sptr> childComponents;
+        const bool recursively = true;
+        assembly->getChildren(childComponents, recursively);
+        for (const auto &childComp : childComponents) {
+          const auto childDetector =
+              boost::dynamic_pointer_cast<const IDetector>(childComp);
+          if (childDetector) {
+            detectorIDs.emplace(childDetector->getID());
+          }
+        }
+      }
```

Comment:
Thanks for the hint! The code indeed gets simplified, even though `Instrument::getDetectorsInBank` doesn't take into account the case where the given component in itself is a detector. Is this intended behavior? The doxygen documentation says *'Fill a vector with all the detectors contained (at any depth) in a named component.'* I would expect 'at any depth' to mean also the component itself.